docs/error_handling.md

```diff
@@ -0,0 +1,200 @@
+
+If you accidentally forget to use one of the error-handling mechanisms,
+you'll get a compilation error along these lines:
+
+```
+required type for x: float
+  but expression 'Raising[(ValueError, KeyError), float](foo_original(x, y))' is of type: Raising[tuple of (ValueError, KeyError), system.float]
+```
+
+Please note that if you have assigned the `Raising` result to a
+variable, the compilation error might happen on a line where you
+attempt to use that variable. To fix the error, please introduce
+error handling as early as possible at the right call-site such
+that no `Raising` variable is created at all.
+
+`noerrors` is another pragma provided for convenience which is
+equivalent to an empty `errors` pragma. The forced error handling
+through the `Raising` type won't be applied.
+
+Both pragmas can be combined with the `nodefects` pragma that
+indicates that the specific proc should be proven to be Defect-free.
+
+The transformation uses a template by default to promote efficiency,
+but if you need to take the address the Raising proc, please add the
+`addressable` pragma that will force the wrapper to be a regular proc.
+
+Finally, `failing` is another pragma provided for convenience which
+is equivalent to `{.errors: (CatchableError).}`.
+
+## The `raising` annotation
+
+The `raising` annotation is the simplest form of error tracking
+similar to the `try` annotation made popular by the [Midori error model](http://joeduffyblog.com/2016/02/07/the-error-model/),
+which is also [proposed for addition in C++](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0709r0.pdf) and [already available in Zig](https://ziglang.org/documentation/master/#try).
+
+It merely marks the locations in the code where errors might be raised
+and strips away the `Raising` type to disarm the compiler checks:
+
+```nim
+proc attachValidator(node: BeaconNode, keyFile: string) {.failing.} =
+  node.addValidator(raising ValidatorPrivKey.init(raising readFile(keyFile))
+```
+
+When applied to a `Result` or an `Option`, `raising` will use the
+`tryGet` API to attempt to obtain the computed value.
+
+## The capital `Try` API
+
+The capital `Try` API is similar to a regular `try` expression
+or a `try` statement. The only difference is that you must provide
+exception handlers for all possible recoverable errors. If you fail
+to do so, the compiler will point out the line in the `try` block
+where an unhandled exception might be raised:
+
+```nim
+proc replaceValues(values: var openarray[string],
+                   replacements: Table[MyEnum, string]) =
+  Try:
+    for v in mitems(values):
+      let
+        enumValue = parseEnum v
+        replacement = replacements[enumValue] # Error here
+      v = replacement
+  except ValueError:
+    echo "Invalid enum value"
+```
+
+The above example will fail to compile with an error indicating
+that `replacements[enumValue]` may fail with an unhandled `KeyError`.
+
+## The `either` expression
```

Comment:
this seems like a small, out-of-scope utility - I wouldn't add until there's significant evidence that it's generally useful in real-world code - using a default after getting an error is a bit weird: the operation failed to calculate a value so now I use a some other value? why didn't the calculation give me that replacement value directly if that makes sense for the type? clearly, this is used when taking "shortcuts" with the type, and seems less legitimate

Reply:
Why would you censor a small utility like this? I find it useful, so let's see how much adoption it will get in practice. it's not just about providing a replacement value - it's also a control-flow construct.

Reply:
because it's meant to become a core concept in the language and imported in every module basically  - you _don't_ want cruft in there from the start - better start every feature at -100 points and work up a case for it, and in this case it's cheap and easy to do "manually"